src/cpu/testers/traffic_gen/gups_gen.cc

```diff
@@ -0,0 +1,444 @@
+
+    // Embed it in a packet
+    PacketPtr pkt = new Packet(req, MemCmd::WriteReq);
+
+    uint8_t* pkt_data = new uint8_t[req->getSize()];
+    pkt->dataDynamic(pkt_data);
+
+    std::memcpy(pkt_data, data, req->getSize());
+
+    return pkt;
+}
+
+void
+GUPSGen::handleResponse(PacketPtr pkt)
+{
+    onTheFlyRequests--;
+    if (pkt->isWrite()) {
+        DPRINTF(GUPSGen, "handleResponse: received a write resp.\n");
+        stats.totalUpdates++;
+        DPRINTF(GUPSGen, "%s: onTheFlyRequests: %d.\n",
+                __func__, onTheFlyRequests);
+
+        stats.totalWrites++;
+        stats.totalBytesWritten += elementSize;
+        stats.totalWriteLat += curTick() - exitTimes[pkt->req];
+
+        exitTimes.erase(pkt->req);
+        delete pkt;
+
+        if ((stats.totalUpdates.value() >= updateLimit) && updateLimit != 0) {
+            exitSimLoop(name() + " has encountered the exit state and will "
+                "terminate the simulation.\n");
+            return;
+        }
+    }
+    else {
+        DPRINTF(GUPSGen, "handleResponse: received a read resp.\n");
+
+        stats.totalReads++;
+        stats.totalBytesRead += elementSize;
+        stats.totalReadLat += curTick() - exitTimes[pkt->req];
+
+        exitTimes.erase(pkt->req);
+
+        responsePool.push(pkt);
+    }
+    if (doneReading && requestPool.empty() && onTheFlyRequests == 0){
+        exitSimLoop(name() + " has encountered the exit state and will "
+                "terminate the simulation.\n");
+        return;
+    }
+    if ((requestPool.size() < reqQueueSize) &&
+        (!doneReading || !responsePool.empty())) {
+        if (!nextCreateEvent.scheduled()) {
+            schedule(nextCreateEvent, nextCycle());
+        }
+    }
+}
+
+void GUPSGen::wakeUp()
+{
+    if (!nextSendEvent.scheduled() && !requestPool.empty())
+    {
+        schedule(nextSendEvent, nextCycle());
+    }
+}
+
+void
+GUPSGen::createNextReq() {
+
+    if (!responsePool.empty()) {
+        PacketPtr pkt = responsePool.front();
+        responsePool.pop();
+
+        uint64_t *updated_value = pkt->getPtr<uint64_t>();
+        *updated_value ^= updateTable[pkt->req];
+        updateTable.erase(pkt->req);
+
+        uint8_t write_data[elementSize];
+        std::memcpy(write_data, updated_value, elementSize);
+        Addr addr = pkt->getAddr();
+        delete pkt;
+
+        PacketPtr new_pkt = getWritePacket(addr, elementSize, write_data);
+        requestPool.push(new_pkt);
+    }
+    else if (!doneReading) {
+        assert (rIndex < 128);
+
+        randomSeeds[rIndex] = (randomSeeds[rIndex] << 1) ^
+                        ((int64_t) randomSeeds[rIndex] < 0 ? POLY : 0);
+        uint64_t value = randomSeeds[rIndex];
+        uint64_t index = randomSeeds[rIndex] & (tableSize - 1);
+        Addr addr = indexToAddr(index);
+        PacketPtr pkt = getReadPacket(addr, sizeof(uint64_t));
+        updateTable[pkt->req] = value;
+        requestPool.push(pkt);
+
+        rIndex++;
```

Comment:
what is this variable???

Reply:
Index for iterating over the randomSeeds array.